### PR TITLE
Checksum reconciled Instances and Bindings

### PR DESCRIPTION
**Pull request by pmorie**
Solves https://github.com/kubernetes-incubator/service-catalog/issues/573 for reprovisions

TLDR:

- Checksum last reconciled ready=true spec
- Controller does not do work if the spec's checksum matches the last reconciled spec

### CODE REVIEW

**Comment by pmorie** [comment]
I didn't realize I wasn't working from master :(

I'll rebase tomorrow.  I was experiencing some local env weirdness today that I didn't have time to sort out, but integration tests should work.

**Comment by pmorie** [comment]
Tested that this has the desired effect on Instance.  Implemented for Binding but not tested yet.

**Comment by pmorie** [comment]
Verified as working on Instances and Bindings

File: pkg/apis/servicecatalog/checksum/unversioned/checksum.go [hunk 1]
```diff
@@ -44,3 +44,23 @@ func InstanceSpecChecksum(spec servicecatalog.InstanceSpec) string {
...
+// - OSBGUID
+func BindingSpecChecksum(spec servicecatalog.BindingSpec) string {
+	specString := ""
+	specString += fmt.Sprintf("instanceRef: %v\n", spec.InstanceRef.Name)
+
+	if spec.Parameters != nil {
+		specString += fmt.Sprintf("parameters:\n\n%v\n\n", string(spec.Parameters.Raw))
+	}
+
+	specString += fmt.Sprintf("osbGuid: %v\n", spec.OSBGUID)
```
Comment (jpeeler) [inline pkg/apis/servicecatalog/checksum/unversioned/checksum.go]: What is the need in producing the hash with anything other than the OSBGUID?

Comment (pmorie) [inline pkg/apis/servicecatalog/checksum/unversioned/checksum.go]: The OSBGuid is static and not expected to change.  In fact, I could probably have omitted it from the checksum.  The fields which are important to checksum are the ones that will mean that the instance needs to be updated at the broker.  Overall I was trying to capture those; it's okay to have fields like the OSBGUID which are static.

**Comment by arschles** [comment]
This LGTM. I have verified that instances are not re-provisioned and bindings are not re-bound as well.

File: pkg/apis/servicecatalog/validation/binding_test.go [hunk 2]
```diff
@@ -48,6 +48,26 @@ func TestValidateBinding(t *testing.T) {
...
+				ObjectMeta: kapi.ObjectMeta{
+					Name:      "test-binding",
+					Namespace: "test-ns",
+				},
+				Spec: servicecatalog.BindingSpec{
+					InstanceRef: v1.LocalObjectReference{
+						Name: "test-instance",
+					},
+					SecretName: "test-secret",
+					Checksum: func() *string {
```
Comment (vaikas-google) [inline pkg/apis/servicecatalog/validation/binding_test.go]: Can this not just be:
Checksum: "boo"

Comment (pmorie) [inline pkg/apis/servicecatalog/validation/binding_test.go]: Nah, it's a pointer to a string :-/

File: pkg/apis/servicecatalog/validation/binding.go [hunk 3]
```diff
@@ -48,15 +52,16 @@ func validateBindingSpec(spec *sc.BindingSpec, fldPath *field.Path) field.ErrorL
...
 	return allErrs
 }
 
 // ValidateBindingUpdate checks that when changing from an older binding to a newer binding is okay.
 func ValidateBindingUpdate(new *sc.Binding, old *sc.Binding) field.ErrorList {
-	allErrs := field.ErrorList{}
-	allErrs = append(allErrs, ValidateBinding(new)...)
-	allErrs = append(allErrs, ValidateBinding(old)...)
-	return allErrs
+	return internalValidateBinding(new, false)
```
Comment (MHBauer) [inline pkg/apis/servicecatalog/validation/binding.go]: @pmorie why only check the new binding? Bindings can't be updated?

**Pull request merged**